src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs

```diff
@@ -0,0 +1,423 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text;
+using ITextBuffer = Microsoft.VisualStudio.Text.ITextBuffer;
+using Timer = System.Threading.Timer;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    internal class DefaultVisualStudioRazorParser : VisualStudioRazorParser, IDisposable
+    {
+        public override event EventHandler<DocumentStructureChangedEventArgs> DocumentStructureChanged;
+
+        // Internal for testing.
+        internal TimeSpan IdleDelay = TimeSpan.FromSeconds(3);
+        internal Timer _idleTimer;
+        internal BackgroundParser _parser;
+
+        private readonly object IdleLock = new object();
+        private readonly ICompletionBroker _completionBroker;
+        private readonly IEnumerable<ICanHasContextChangedListener> _parserContextChangedListeners;
```

Comment:
We will need a more serious name than this, even if it's temporary. #NoFunRule

Consider this a blocker. `IContextChangedListener` would be fine